### PR TITLE
Detect and handle reorganizations

### PR DESCRIPTION
**Pull request by omerfirmak**
Depends on #735

### CODE REVIEW

File: blockchain/blockchain.go [hunk 1]
```diff
@@ -281,25 +285,22 @@ func (b *Blockchain) verifyBlock(txn db.Transaction, block *core.Block) error {
 		return err
 	}
 
+	expectedBlockNumber := uint64(0)
+	expectedParentHash := &felt.Zero
+
 	head, err := b.head(txn)
 	if err != nil && !errors.Is(err, db.ErrKeyNotFound) {
```
Comment (wojciechos) [inline blockchain/blockchain.go:292]: Maybe we could make it a little bit shorter like:
```
if err == nil {
    expectedBlockNumber = head.Number + 1
    expectedParentHash = head.Hash
} else if !errors.Is(err, db.ErrKeyNotFound) {
    return err
}
```

Comment (omerfirmak) [reply blockchain/blockchain.go:292]: yeap, looks neater. done

File: core/state.go [hunk 2]
```diff
@@ -406,15 +417,18 @@ func calculateContractCommitment(storageRoot, classHash, nonce *felt.Felt) *felt
...
 	if err != nil {
 		return err
 	}
 
 	for _, declaredClass := range declaredClasses {
 		// https://docs.starknet.io/documentation/starknet_versions/upcoming_versions/#commitment
-		leafValue := crypto.Poseidon(leafVersion, declaredClass.CompiledClassHash)
+		leafValue := &felt.Zero
+		if !revert {
+			leafValue = crypto.Poseidon(leafVersion, declaredClass.CompiledClassHash)
```
Comment (rianhughes) [inline core/state.go:430]: Is setting the value to zero equivalent to deletion (in terms of taking up memory, querying the value of a key, etc)?

Comment (rianhughes) [inline core/state.go:430]: Does this set the leafValue to zero rather than whatever the old value was? Eg if we are reverting a declared class that had changed it's value from say '1' to '2', would this not revert it to '0' instead of '1'?

File: core/state.go [hunk 3]
```diff
@@ -437,3 +451,135 @@ func (s *State) ContractIsAlreadyDeployedAt(addr *felt.Felt, blockNumber uint64)
...
+	return s.expectRoot(update.OldRoot)
+}
+
+func (s *State) purgeContract(addr *felt.Felt) error {
+	contract, err := NewContract(addr, s.txn)
+	if err != nil {
+		return err
+	}
+
+	state, storageCloser, err := s.storage()
```
Comment (kirugan) [inline core/state.go:494]: is it ok if we don't call `storageCloser` in case of failure methods below?

File: core/state.go [hunk 4]
```diff
@@ -437,3 +451,135 @@ func (s *State) ContractIsAlreadyDeployedAt(addr *felt.Felt, blockNumber uint64)
 	}
 	return deployedAt <= blockNumber, nil
 }
+
+func (s *State) Revert(blockNumber uint64, update *StateUpdate) error {
+	err := s.expectRoot(update.NewRoot)
+	if err != nil {
+		return err
+	}
```
Comment (rianhughes) [inline core/state.go:459]:
```suggestion
	if err := s.expectRoot(update.NewRoot); err != nil {
		return err
	}
```

File: core/state.go (hunk 2 above)
Comment (omerfirmak) [reply core/state.go:430]: Leafs in the classes trie are immutable, so they can only be set once to a non-zero value. So the old value is always zero.

File: core/state.go (hunk 3 above)
Comment (omerfirmak) [reply core/state.go:494]: No it is not:) Will fix, thanks!

File: core/state.go (hunk 4 above)
Comment (omerfirmak) [reply core/state.go:459]: This is actually intentional, it let's us reuse err variable throughout the entire functional call.

File: sync/sync_test.go [hunk 5]
```diff
@@ -146,3 +146,45 @@ func TestSyncBlocks(t *testing.T) {
...
+		synchronizer := New(bc, mainGw, utils.NewNopZapLogger())
+		ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+		require.NoError(t, synchronizer.Run(ctx))
+		cancel()
+
+		t.Run("head is now MAINNET", func(t *testing.T) {
+			head, err := bc.HeadsHeader()
+			require.NoError(t, err)
+			require.Equal(t, utils.HexToFelt(t, "0x4e1f77f39545afe866ac151ac908bd1a347a2a8a7d58bef1276db4f06fdf2f6"), head.Hash)
+		})
```
Comment (IronGauntlets) [inline sync/sync_test.go:188]: lines 179 -182 should be a part of the sub-test.

However, if the earlier subtest tests are removed having one subtest makes no sense.

**Review by IronGauntlets — CHANGES_REQUESTED** [review]

File: sync/sync.go [hunk 6]
```diff
@@ -140,8 +140,15 @@ func (s *Synchronizer) verifierTask(ctx context.Context, block *core.Block, stat
 			}
 			err = s.Blockchain.Store(block, stateUpdate, newClasses)
 			if err != nil {
-				s.log.Warnw("Failed storing Block", "number", block.Number,
-					"hash", block.Hash.ShortString(), "err", err.Error())
+				if errors.Is(err, blockchain.ErrParentDoesNotMatchHead) {
+					// revert the head and restart the sync process, hoping that the reorg is not deep
+					// if the reorg is deeper, we will end up here again and again until we fully revert reorged
+					// blocks
```
[the remark this answers is not in the document]
Comment (omerfirmak) [reply sync/sync.go:146]: Yeap, it is not ideal but it is the easiest to implement. Given that the reorgs are not common and Synchronizer is bound to get refactored after P2P, I didnt want to unnecessarily complicate it further.

File: core/state.go [hunk 7]
```diff
@@ -243,13 +251,6 @@ func (s *State) updateContracts(blockNumber uint64, diff *StateDiff, logChanges
 		return err
 	}
 
-	// register deployed contracts
-	for _, contract := range diff.DeployedContracts {
-		if err := s.putNewContract(stateTrie, contract.Address, contract.ClassHash, blockNumber); err != nil {
-			return err
-		}
-	}
-
```
Comment (omerfirmak) [inline core/state.go:242]: Why was this removed?

File: core/contract.go [hunk 8]
```diff
@@ -98,6 +98,22 @@ type Contract struct {
 	txn db.Transaction
 }
 
+// Purge nukes the contract instance by deleting all related data from storage
```
[the remark this answers is not in the document]
Comment (kirugan) [reply core/contract.go]: repharsed it with `Purge eliminates the contract instance, deleting all associated data from storage`

File: core/state.go [hunk 9]
```diff
@@ -186,16 +186,26 @@ func (s *State) globalTrie(bucket db.Bucket, newTrie trie.NewTrieFunc) (*trie.Tr
 	return gTrie, closer, nil
 }
 
+func (s *State) expectRoot(root *felt.Felt) error {
```
[the remark this answers is not in the document]
Comment (kirugan) [reply core/state.go]: renamed to `verifyStateUpdateRoot(...)`

File: core/state.go [hunk 10]
```diff
@@ -206,40 +216,36 @@ func (s *State) Update(blockNumber uint64, update *StateUpdate, declaredClasses
 		}
 	}
 
-	if err = s.updateDeclaredClasses(update.StateDiff.DeclaredV1Classes); err != nil {
+	if err = s.updateDeclaredClasses(update.StateDiff.DeclaredV1Classes, false); err != nil {
```
[the remark this answers is not in the document]
Comment (kirugan) [reply core/state.go]: renamed

File: sync/sync_test.go [hunk 11]
```diff
@@ -146,3 +146,45 @@ func TestSyncBlocks(t *testing.T) {
...
+	testDB := pebble.NewMemTest()
+
+	t.Run("sync to integration for 2 blocks", func(t *testing.T) {
+		bc := blockchain.New(testDB, utils.INTEGRATION, utils.NewNopZapLogger())
+		synchronizer := New(bc, integGw, utils.NewNopZapLogger())
+
+		ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+		require.NoError(t, synchronizer.Run(ctx))
+		cancel()
+	})
```
[the remark this answers is not in the document]
Comment (kirugan) [reply sync/sync_test.go]: moved outside subtest, the only question I have: how do we specify here that it's exactly/at least 2 blocks? :)

File: sync/sync_test.go [hunk 12]
```diff
@@ -146,3 +146,45 @@ func TestSyncBlocks(t *testing.T) {
 		testBlockchain(t, bc)
 	})
 }
+
+func TestReorg(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (kirugan) [reply sync/sync_test.go:150]: How should we migrate existing databases when there isn't a designated key yet? We could either store the used network flag in the provided database path, or double-check that the database data aligns with the network data.

File: core/state.go (hunk 7 above)
Comment (omerfirmak) [reply core/state.go:242]: something must have gone wrong while rebasing

File: core/state.go [hunk 13]
```diff
@@ -210,24 +220,26 @@ func (s *State) Update(blockNumber uint64, update *StateUpdate, declaredClasses
...
 		return err
 	}
 
-	if err = s.updateContracts(blockNumber, update.StateDiff); err != nil {
-		return err
+	// register deployed contracts
+	for _, contract := range update.StateDiff.DeployedContracts {
+		// todo fix after rebase
```
Comment (omerfirmak) [inline core/state.go]: Please keep the individual commits clean. ie builds fine and passes all tests

**Comment by omerfirmak** [comment]
Let me rebase the original PR and apply your changes on top of it. @kirugan

File: core/state.go [hunk 14]
```diff
@@ -437,3 +451,135 @@ func (s *State) ContractIsAlreadyDeployedAt(addr *felt.Felt, blockNumber uint64)
...
 	return deployedAt <= blockNumber, nil
 }
+
+func (s *State) Revert(blockNumber uint64, update *StateUpdate) error {
+	err := s.expectRoot(update.NewRoot)
+	if err != nil {
+		return err
+	}
+
+	// todo: remove classes from storage
```
[the remark this answers is not in the document]
Comment (kirugan) [reply core/state.go]: Added deletion for declared classes (v0, v1). Is class replacement just an upgrade for classes or something else?

[the remark this answers is not in the document]
Comment (omerfirmak) [reply core/state.go]: Replacement means just switching the underlying implementation of a contract at a given address. It doesnt have anything to do with class storage. We already handle class replacements both in forward and reverse state updates.

> Added deletion for declared classes (v0, v1).

can you add a testcase for it as well?

[the remark this answers is not in the document]
Comment (omerfirmak) [reply core/state.go]: > can you add a testcase for it as well?

there should a test here that declares a class and then reverts that block

https://github.com/NethermindEth/juno/pull/707/files#diff-1b239d70750c2dc052c0f84a84b2a0b8291d82a04d67bbd2779c7e4d84bee13dR367-R460

File: core/state_test.go [hunk 15]
```diff
@@ -363,3 +364,97 @@ func TestClass(t *testing.T) {
 	assert.Zero(t, gotCairo0Class.At)
 	assert.Equal(t, cairo0Class, gotCairo0Class.Class)
 }
+
+func TestRevert(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (omerfirmak) [reply core/state_test.go:368]: What do you mean exactly?

File: core/state.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (kirugan) [reply core/state.go]: @omerfirmak could you please check added tests 🙏

File: core/state_test.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (IronGauntlets) [reply core/state_test.go:368]: A testcase that should check the values in the storage diffs have been reverted correctly. After reverting the state should return old storage values and not the ones contained in the state update's storage diff.

We have testcase for replaced class, declared class and storage root.

[the remark this answers is not in the document]
Comment (omerfirmak) [reply core/state_test.go:368]: Since state root is checked after revert is done, individual storage values are implicitly verified as well

File: sync/sync_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (IronGauntlets) [reply sync/sync_test.go:150]: I am resolving this conversation as this doesn't require any change.

File: core/state_test.go [hunk 16]
```diff
@@ -426,6 +426,27 @@ func TestRevert(t *testing.T) {
...
+				DeclaredV1Classes: []core.DeclaredV1Class{
+					{
+						ClassHash:         utils.HexToFelt(t, "0xDEADBEEF2"),
+						CompiledClassHash: utils.HexToFelt(t, "0xDEADBEEF3"),
+					},
+				},
+			},
+		}
+
+		require.NoError(t, state.Update(2, declaredClassesStateUpdate, nil))
```
[the remark this answers is not in the document]
Comment (IronGauntlets) [reply core/state_test.go]: Done, please check latest commit

File: sync/sync_test.go (hunk 5 above)
Comment (IronGauntlets) [reply sync/sync_test.go]: see e96d7e2faa63ddea1caec276ddd3f7fa33135ed2

**Review by IronGauntlets — APPROVED** [review]

**Review by kirugan — APPROVED** [review]

**Review by kirugan — APPROVED** [review]
